---
title: "`r MotifPeeker::report_header()`"
date: "`r format(Sys.Date(), '%b-%d-%Y')`"
output: 
    html_document: 
        toc: yes
        toc_float: yes
        code_folding: hide 
        number_sections: true
        css: custom.css
params:
    peak_files:
        value: NULL
    reference_index:
        value: 1
    alignment_files:
        value: NULL
    exp_labels:
        value: NULL
    exp_type:
        value: NULL
    genome_build:
        value: NULL
    motif_files:
        value: NULL
    motif_labels:
        value: NULL
    cell_counts:
        value: NULL
    denovo_motif_discovery:
        value: NULL
    denovo_motifs:
        value: 3
    motif_db:
        value: NULL
    download_buttons:
        value: NULL
    output_dir:
        value: NULL
    use_cache:
        value: NULL
    ncpus:
        value: NULL
    debug:
        value: FALSE
    verbose:
        value: TRUE
---

```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                    message = params$debug,
                    warning = params$debug,
                    cache = FALSE, 
                    error = params$debug,
                    results = "asis")

### Parse input ###
genome_build <- check_genome_build(params$genome_build)
result_len <- length(params$peak_files)
if (is.null(params$exp_type)) {
    exp_types <- rep("unknown", result_len)
} else {
    exp_types <- params$exp_type
}
peak_files_encode <- Vectorize(check_ENCODE, "encode_id")(
    params$peak_files, expect_format = c("narrowPeak", "bed"))
alignment_files_encode <- Vectorize(check_ENCODE, "encode_id")(
    params$alignment_files, expect_format = c("bam", "bed"))
result <- list(
    peaks = Vectorize(read_peak_file, "peak_file")(peak_files_encode),
    alignments = lapply(alignment_files_encode, Rsamtools::BamFile),
    exp_labels = params$exp_labels,
    exp_type = Vectorize(format_exptype,
                        "exp_type")(exp_types)
)
motif_files_jaspar <- Vectorize(check_JASPAR, "motif_id")(params$motif_files)
user_motifs <- list(
    motifs = Vectorize(read_motif_file, "motif_file")(motif_files_jaspar),
    motif_labels = params$motif_labels
)

## Flags for optional plots
alignment_metrics <- ifelse((length(result$alignments) == 0), FALSE, TRUE)
cellcount_metrics <- ifelse((length(params$cell_counts) == 0), FALSE, TRUE)

## Misc
ex_emo <- emoji::emoji("exclamation")

## Check motif_db
if (params$denovo_motif_discovery && is.null(params$motif_db)) {
    motif_db <- get_JASPARCORE()
} else {
    motif_db <- params$motif_db
}

### General Metrics ###
if (alignment_metrics) {
    ## FRiP scores
    result$frip <- Vectorize(calc_frip, c("read_file",
                                "peak_file"))(result$alignments, result$peaks)
    frip_df <- data.frame(exp_type = result$exp_type,
                            frip = result$frip,
                            exp_label = result$exp_labels)
    rownames(frip_df) <- seq_len(nrow(frip_df))
}

```

```{r setup_misc, include = FALSE}
## Junk plot to add Plotly JS
plotly::plot_ly(x = c(1,2,3), type = "histogram")
```

[`MotifPeeker`](https://github.com/neurogenomics/MotifPeeker) compares different
epigenomic datasets using motif enrichment as the key metric.

# Summary {-}
## Table of Contents {-}
The report consists of the following sections:  

1. [General Metrics](#general-metrics): Overview of general metrics related to
peaks in the datasets (FRiP scores, peak widths, and motif-summit distances).

2. [Known Motif Enrichment Analysis](#motif-enrichment-analysis): Statistics on
the frequency of motifs enriched in the datasets. Also compares enriched motifs
between common and unique peaks identified in the datasets.

3. [De-Novo Motif Enrichment Analysis](#denovo-motif-analysis): Statistics on
de-novo motifs discovered between common and unique peaks identified in the
datasets. Also checks for similarity between motifs in each set and finds the
closest known motif in the [JASPAR database](https://jaspar.uio.no/downloads/)
(or the supplied database).

## Input Datasets {-}
Experimental dataset labels used:  
```{r echo = FALSE}
for (i in seq_along(result$peaks)) {
    cat(paste0("\n- **Dataset ", i, "**: ", result$exp_labels[i], "  \n"))
    cat(paste0("  - **Experiment Type**: ", result$exp_type[i], "  \n"))
    if (!is.null(params$cell_counts)) {
        cat(paste0("  - **Cell Count**: ", params$cell_counts[i], "  \n"))
    }
}
cat("\n- **Reference dataset**: ", result$exp_labels[params$reference_index], 
    paste0("(Index: ", params$reference_index, ")  \n"))
```

User-provided motifs used:  

```{r echo = FALSE}
if (length(user_motifs$motifs) == 0) cat("- **None**  \n")
for (i in seq_along(user_motifs$motifs)) {
    cat(paste0("- **Motif ", i, "**: ", user_motifs$motifs[[i]]@name, "  \n"))
}
```

## Command {-}
`MotifPeeker` report was generated with the following command:
```{r report_command}
cat(report_command(params))
```

## Summary Plot {-}
Lorem ipusm


# General Metrics {.tabset #general-metrics}

## FRiP Score
**Fraction of Reads in Peaks (FRiP)** is defined as the ratio of reads
overlapping peaks to the total number of reads in the experiment. Higher FRiP
score indicates higher enrichment of reads in peaks.  
$$\text{FRiP (Fraction of Reads in Peaks)} = 
\frac{\text{Reads in Peaks}}{\text{Total Reads}}$$
```{r frip_exp_plot}
if (!alignment_metrics) {
    cat("\n", ex_emo,
    " *Skipping FRiP calculation as alignment files are not provided.*",
        "  \n")
} else {
    if (length(result$alignments) != length(result$peaks)) {
        stp_msg <- "Number of alignment files and peak files do not match."
        stopper(stp_msg)
    }
    cat("\n### By Experiment Type {-}  \n")
    frip_exp_plot <- frip_df |>
        ggplot2::ggplot(aes(x = exp_type, y = frip, fill = exp_type,
                            text = paste0(
                                "<b>Experiment Label</b>: ", exp_label, "<br>",
                                "<b>Experiment Type</b>: ", exp_type, "<br>",
                                "<b>FRiP Score</b>: ", round(frip, 4)))) +
            geom_boxplot(outlier.shape = NA) +
            geom_jitter() +
            labs(
              x = "Experiment Type",
              y = "FRiP Score",
              fill = "Experiment Type"
            ) +
        scale_fill_viridis(begin = 0.1, end = 0.5, discrete = TRUE,
                           option = "A")
    frip_exp_plotly <- plotly::ggplotly(frip_exp_plot, tooltip = "text")
    print(htmltools::tagList(frip_exp_plotly))
}
```

```{r frip_exp_dt, include = alignment_metrics}
if (alignment_metrics) {
    frip_df |>
        group_by(exp_type) |>
        rename("Experiment Type" = exp_type) |>
        summarise(
          "Mean score" = round(mean(frip), 3),
          "Median Score" = round(median(frip), 3),
          "Standard Deviation" = round(sd(frip), 3)
        ) |>
    print_DT()
}
```
```{r frip_individual_plot, include = alignment_metrics}
if (alignment_metrics) {
    cat("\n### By Individual Dataset {-}  \n")
    
    frip_individual_plot <- frip_df |>
        ggplot(aes(x = exp_label, y = frip, fill = exp_type, text = paste0(
                                "<b>Experiment Label</b>: ", exp_label, "<br>",
                                "<b>Experiment Type</b>: ", exp_type, "<br>",
                                "<b>FRiP Score</b>: ", round(frip, 4)))) +
            theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
            geom_col() +
            labs(
              x = "Experiment Label",
              y = "FRiP Score",
              fill = "Experiment Type"
            ) +
        scale_fill_viridis(begin = 0.1, end = 0.5, discrete = TRUE,
                           option = "A")
    frip_individual_plotly <- plotly::ggplotly(frip_individual_plot,
                                               tooltip = "text")
    print(htmltools::tagList(frip_individual_plotly))
}
```
```{r frip_individual_dt, include = alignment_metrics}
if (alignment_metrics) {
    frip_df$read_count <- lapply(result$alignments, function(x) {
        Rsamtools::countBam(x)$records
    }) |> unlist()
    frip_individual_df <- frip_df |>
        mutate(frip = round(frip, 3)) |>
        select(exp_label, exp_type, read_count, frip) |>
        rename(
          "Experiment Label" = exp_label,
          "Experiment Type" = exp_type,
          "FRiP Score" = frip,
          "Read Count" = read_count
        )
    
    if (length(params$cell_counts) > 0) {
        frip_individual_df$cell_count <- params$cell_counts
        frip_individual_df <- frip_individual_df |>
            rename(
              "Cell Count" = cell_count
            )
    }
    frip_individual_df |>
        print_DT()
}
```
```{r frip_cellcount_plot, include = alignment_metrics}
if (alignment_metrics) {
    if (!cellcount_metrics) {
        cat("\n", ex_emo,
        " *Skipping cell count plot as cell counts are not provided.*\n")
    } else {
        cat("\n### By Individual Dataset {-}  \n")
        
        frip_df$cell_count <- as.numeric(params$cell_counts)
        frip_df_cellcount <- frip_df |>
            group_by(cell_count, exp_type) %>%
                summarise(
                    frip = mean(frip),
                    .groups = "drop"
                )
            
        frip_cellcount_plot <- frip_df |>
            ggplot(aes(
                x = cell_count,
                y = frip,
                shape = exp_type,
                color = exp_type,
                text = paste0("<b>Experiment Type</b>: ", exp_type, "<br>",
                                "<b>FRiP Score</b>: ", round(frip, 4),
                                "<br><b>Cell Count</b>: ", cell_count))) +
            geom_point() +
            geom_line(data = frip_df_cellcount) +
            labs(
                x = "Cell Count (log10 scale)",
                y = "FRiP Score",
                fill = "Experiment Type",
                shape = "Experiment Type",
                color = "Experiment Type"
            ) +
            scale_x_continuous(trans="log10") +
            scale_color_viridis(begin = 0.1, end = 0.5, discrete = TRUE,
                           option = "A")
        
        frip_cellcount_plot <- plotly::ggplotly(frip_cellcount_plot,
                                                tooltip = "text")
        
        print(htmltools::tagList(frip_cellcount_plot))
    }
}
```


# Session Info {-}
<details> 
```{r Session Info, results='markup'}
utils::sessionInfo()
```
</details> 
<hr>
