---
title: "`r MotifPeeker::report_header()`"
date: "`r format(Sys.Date(), '%b-%d-%Y')`"
output: 
    html_document: 
        toc: yes
        toc_float: yes
        code_folding: hide 
        number_sections: true
        css: custom.css
params:
    peak_files:
        value: NULL
    reference_index:
        value: 1
    alignment_files:
        value: NULL
    exp_labels:
        value: NULL
    exp_type:
        value: NULL
    genome_build:
        value: NULL
    motif_files:
        value: NULL
    motif_labels:
        value: NULL
    cell_counts:
        value: NULL
    denovo_motif_discovery:
        value: NULL
    denovo_motifs:
        value: 3
    motif_db:
        value: NULL
    trim_seq_width:
        value: NULL
    download_buttons:
        value: NULL
    meme_path:
        value: NULL
    output_dir:
        value: NULL
    save_runfiles:
        value: NULL
    use_cache:
        value: NULL
    workers:
        value: NULL
    debug:
        value: FALSE
    verbose:
        value: TRUE
---

```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                    message = params$debug,
                    warning = params$debug,
                    cache = FALSE, 
                    error = params$debug,
                    results = "asis")

### Parse input ###
genome_build <- check_genome_build(params$genome_build)
result_len <- length(params$peak_files)
motif_len <- length(params$motif_files)
if (is.null(params$exp_type)) {
    exp_types <- rep("unknown", result_len)
} else {
    exp_types <- params$exp_type
}
peak_files_encode <- Vectorize(check_ENCODE, "encode_id")(
    params$peak_files, expect_format = c("narrowPeak", "bed"))
alignment_files_encode <- Vectorize(check_ENCODE, "encode_id")(
    params$alignment_files, expect_format = "bam")
result <- list(
    peaks = Vectorize(read_peak_file, "peak_file")(peak_files_encode),
    alignments = lapply(alignment_files_encode, Rsamtools::BamFile),
    exp_labels = params$exp_labels,
    exp_type = unname(Vectorize(format_exptype,
                        "exp_type")(exp_types))
)
motif_files_jaspar <- if (motif_len > 1) {
        Vectorize(check_JASPAR, "motif_id")(params$motif_files)
    } else {
        check_JASPAR(params$motif_files)
    }
user_motifs <- list(
    motifs = if (motif_len > 1) {
            Vectorize(read_motif_file, "motif_file")(motif_files_jaspar)
        } else if (motif_len == 1) {
            read_motif_file(motif_files_jaspar)
        },
    motif_labels = params$motif_labels
)
result$alignments <- unname(result$alignments)
result$peaks <- unname(result$peaks)
user_motifs$motifs <- unname(user_motifs$motifs)

out_dir_extra <- ifelse(params$save_runfiles, params$out_dir,
                       file.path(tempdir(), paste0("MotifPeeker_temp_",
                                                   random_string(5))))

## Flags for optional plots
alignment_metrics <- ifelse((length(result$alignments) == 0), FALSE, TRUE)
cellcount_metrics <- ifelse((length(params$cell_counts) == 0), FALSE, TRUE)
user_motif_metrics <- ifelse((length(user_motifs$motifs) == 0), FALSE, TRUE)

## Misc
ex_emo <- ifelse(requireNamespace("emoji", quietly = TRUE),
                emoji::emoji("exclamation"), "!!")

## Check motif_db
motif_db <- ifelse(params$denovo_motif_discovery && is.null(params$motif_db),
                    get_JASPARCORE(), params$motif_db)

### General Metrics ###
## Peak Counts
result$peak_count <- vapply(
        result$peaks, function(x) length(x), integer(1)
    )
if (alignment_metrics) {
    ## Read Counts
    result$read_count <- vapply(result$alignments, function(x) {
        Rsamtools::countBam(x)$records
    }, integer(1))
    
    ## FRiP scores
    result$frip <- Vectorize(calc_frip, c("read_file",
                                "peak_file"))(result$alignments, result$peaks)
    frip_df <- data.frame(exp_type = result$exp_type,
                            frip = result$frip,
                            exp_label = result$exp_labels)
}
    
## Peak Widths
result$peak_widths <- lapply(result$peaks, function(x){
    x@ranges@width
})
peak_width_len <- vapply(result$peak_widths, length, integer(1))
peak_width_df <- 
    data.frame(exp_label = rep(result$exp_labels, peak_width_len),
                exp_type = rep(result$exp_type, peak_width_len),
                peak_width = unlist(result$peak_widths)
                )

## Motif-Summit Distances
if (length(user_motifs$motifs) == 1 ) user_motifs$motifs <-
    list(user_motifs$motifs)
motif_summit_dist_df <- get_df_distances(
        result, user_motifs, genome_build, out_dir_extra, params$workers,
        params$verbose
    )

### Known-motif Analysis ###
comparison_indices <- setdiff(seq_along(result$peaks), params$reference_index)
segregated_peaks <- lapply(comparison_indices, function(x) {
    segregate_seqs(result$peaks[[params$reference_index]], result$peaks[[x]])
})

## Calculate enrichment for segregated peaks
## 1. All peaks
peak_combinations <- rep(result$peaks, each = motif_len)
exp_label_combinations <- rep(result$exp_labels, each = motif_len)
exp_type_combinations <- rep(result$exp_type, each = motif_len)
motif_combinations <- rep(user_motifs$motifs, result_len)

enrichment_df_all <- bpapply(
    seq_len(result_len * motif_len),
    function(i) {
        peak <- peak_combinations[[i]]
        motif <- motif_combinations[[i]]
        res <- MotifPeeker::motif_enrichment(
            peak, motif,
            genome_build = genome_build,
            out_dir = file.path(out_dir_extra, "ame", i)
            )
        list(
            exp_label = exp_label_combinations[i],
            exp_type = exp_type_combinations[i],
            group1 = "all",
            group2 = ifelse(exp_label_combinations[i] ==
                                result$exp_labels[params$reference_index],
                            "reference", "comparison"),
            motif_indice = rep(seq_len(motif_len), result_len)[i],
            count_enriched = res$tp[1],
            count_nonenriched = rep(result$peak_count, each = motif_len)[i] -
                res$tp[1],
            perc_enriched = res$tp[2],
            perc_nonenriched = 100 - res$tp[2]
        )
    },
    workers = params$workers, verbose = params$verbose) %>%
    purrr::map_df(as.data.frame)

## 2. Segregated peaks
## Output: Peak1 common_seq1 - Motif 1
##         Peak1 common_seq1 - Motif 2
##         ...
##         Peak1 common_seq2 - Motif 1
##         ...
##         Peak2 common_seq1 - Motif 1
##         ...
peak_combinations2 <- lapply(seq_along(comparison_indices), function(i) {
    rep(segregated_peaks[[i]], each = motif_len)
}) %>% unlist
exp_label_combinations2 <- lapply(comparison_indices, function(i) {
    rep(rep(result$exp_labels[i], 4), each = motif_len)
}) %>% unlist
exp_type_combinations2 <- lapply(comparison_indices, function(i) {
    rep(rep(result$exp_type[i], 4), each = motif_len)
}) %>% unlist
motif_combinations2 <- lapply(comparison_indices, function(i) {
    rep(seq_along(user_motifs$motifs), 4) # Indices
}) %>% unlist %>% unname
group1_labels <- rep(c("common", "unique"), each = 2 * motif_len,
                    length(comparison_indices))
group2_labels <- rep(c("reference", "comparison"), each = motif_len,
                    length(comparison_indices) * 2)
enrichment_df_seg <- bpapply(
    seq_len((result_len - 1) * 4 * motif_len),
    function(i) {
        peak <- peak_combinations2[[i]]
        motif <- user_motifs$motifs[motif_combinations2[[i]]]
        res <- MotifPeeker::motif_enrichment(
            peak, motif,
            genome_build = genome_build,
            out_dir = file.path(out_dir_extra, "ame_segregated", i)
            )
        list(
            exp_label = exp_label_combinations2[i],
            exp_type = exp_type_combinations2[i],
            group1 = group1_labels[i],
            group2 = group2_labels[i],
            motif_indice = motif_combinations2[i],
            count_enriched = res$tp[1],
            count_nonenriched = length(peak) - res$tp[1],
            perc_enriched = res$tp[2],
            perc_nonenriched = 100 - res$tp[2]
        )
    },
    workers = params$workers, verbose = params$verbose) %>%
    purrr::map_df(as.data.frame)

enrichment_df <- rbind(enrichment_df_all, enrichment_df_seg)

```

```{r setup_misc, include = FALSE}
## Junk plot to add Plotly JS
plotly::plot_ly(x = c(1,2,3), type = "histogram")
```

[`MotifPeeker`](https://github.com/neurogenomics/MotifPeeker) compares different
epigenomic datasets using motif enrichment as the key metric.

# Summary {-}
## Table of Contents {-}
The report consists of the following sections:  

1. [General Metrics](#general-metrics): Overview of general metrics related to
peaks in the datasets (FRiP scores, peak widths, and motif-summit distances).

2. [Known Motif Enrichment Analysis](#motif-enrichment-analysis): Statistics on
the frequency of motifs enriched in the datasets. Also compares enriched motifs
between common and unique peaks identified in the datasets.

3. [De-Novo Motif Enrichment Analysis](#denovo-motif-analysis): Statistics on
de-novo motifs discovered between common and unique peaks identified in the
datasets. Also checks for similarity between motifs in each set and finds the
closest known motif in the [JASPAR database](https://jaspar.uio.no/downloads/)
(or the supplied database).

## Input Datasets {-}
Experimental dataset labels used:  
```{r echo = FALSE}
for (i in seq_along(result$peaks)) {
    cat(paste0("\n- **Dataset ", i, "**: ", result$exp_labels[i], "  \n"))
    cat(paste0("  - **Experiment Type**: ", result$exp_type[i], "  \n"))
    if (!is.null(params$cell_counts)) {
        cat(paste0("  - **Cell Count**: ", params$cell_counts[i], "  \n"))
    }
}
cat("\n- **Reference dataset**: ", result$exp_labels[params$reference_index], 
    paste0("(Index: ", params$reference_index, ")  \n"))
```

User-provided motifs used:  

```{r echo = FALSE}
if (length(user_motifs$motifs) == 0) cat("- **None**  \n")
for (i in seq_along(user_motifs$motifs)) {
    cat(paste0("- **Motif ", i, "**: ", user_motifs$motifs[[i]]@name, "  \n"))
}
```

## Command {-}
`MotifPeeker` report was generated with the following command:
```{r report_command}
cat(report_command(params))
```

## Summary Plot {-}
Lorem ipusm

<hr>

# General Metrics {.tabset #general-metrics}

## FRiP Score {-}
**Fraction of Reads in Peaks (FRiP)** is defined as the ratio of reads
overlapping peaks to the total number of reads in the experiment. Higher FRiP
score indicates higher enrichment of reads in peaks.  
$$\text{FRiP (Fraction of Reads in Peaks)} = 
\frac{\text{Reads in Peaks}}{\text{Total Reads}}$$
```{r frip_exp_plot}
if (!alignment_metrics) {
    cat("\n", ex_emo,
    " *Skipping FRiP calculation as alignment files are not provided.*",
        "  \n")
} else {
    if (length(result$alignments) != length(result$peaks)) {
        stp_msg <- "Number of alignment files and peak files do not match."
        stopper(stp_msg)
    }
    cat("\n### By Experiment Type {- .unlisted}  \n")
    frip_exp_plot <- frip_df |>
        ggplot2::ggplot(aes(x = exp_type, y = frip, fill = exp_type,
                            text = paste0(
                                "<b>Experiment Label</b>: ", exp_label, "<br>",
                                "<b>Experiment Type</b>: ", exp_type, "<br>",
                                "<b>FRiP Score</b>: ", round(frip, 4)))) +
            geom_boxplot(outlier.shape = NA) +
            geom_jitter() +
            labs(
              x = "Experiment Type",
              y = "FRiP Score",
              fill = "Experiment Type"
            ) +
        scale_fill_viridis(begin = 0.15, end = 0.6, discrete = TRUE,
                           option = "A")
    
    print(to_plotly(frip_exp_plot))
}
```

```{r frip_exp_dt, include = alignment_metrics}
if (alignment_metrics) {
    frip_df %>%
        group_by(exp_type) %>%
        rename("Experiment Type" = exp_type) %>%
        summarise(
          "Mean score" = round(mean(frip), 3),
          "Median Score" = round(median(frip), 3),
          "Standard Deviation" = round(sd(frip), 3)
        ) %>%
        arrange(`Experiment Type`) %>%
    print_DT()
}
```
```{r frip_individual_plot, include = alignment_metrics}
if (alignment_metrics) {
    cat("\n### By Individual Dataset {- .unlisted}  \n")
    
    frip_individual_plot <- frip_df %>%
        ggplot(aes(x = exp_label, y = frip, fill = exp_type, text = paste0(
                                "<b>Experiment Label</b>: ", exp_label, "<br>",
                                "<b>Experiment Type</b>: ", exp_type, "<br>",
                                "<b>FRiP Score</b>: ", round(frip, 4)))) +
            theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
            geom_col() +
            labs(
              x = "Experiment Label",
              y = "FRiP Score",
              fill = "Experiment Type"
            ) +
        scale_fill_viridis(begin = 0.15, end = 0.6, discrete = TRUE,
                           option = "A")
    
    print(to_plotly(frip_individual_plot))
}
```
```{r frip_individual_dt, include = alignment_metrics}
if (alignment_metrics) {
    frip_df$read_count <- result$read_count
    frip_individual_df <- frip_df %>%
        mutate(frip = round(frip, 3)) %>%
        select(exp_label, exp_type, read_count, frip) %>%
        rename(
          "Experiment Label" = exp_label,
          "Experiment Type" = exp_type,
          "FRiP Score" = frip,
          "Read Count" = read_count
        ) %>%
        arrange(`Experiment Type`, `Experiment Label`)
    
    if (length(params$cell_counts) > 0) {
        frip_individual_df$cell_count <- params$cell_counts
        frip_individual_df <- frip_individual_df %>%
            rename(
              "Cell Count" = cell_count
            )
    }
    frip_individual_df %>%
        print_DT()
}
```
```{r frip_cellcount_plot, include = alignment_metrics}
if (alignment_metrics) {
    if (!cellcount_metrics) {
        cat("\n", ex_emo,
        " *Skipping cell count plot as cell counts are not provided.*\n")
    } else {
        cat("\n### By Cell Count {- .unlisted}  \n")
        
        frip_df$cell_count <- as.numeric(params$cell_counts)
        frip_df_cellcount <- frip_df |>
            group_by(cell_count, exp_type) %>%
                summarise(
                    frip = mean(frip),
                    .groups = "drop"
                )
            
        frip_cellcount_plot <- frip_df |>
            ggplot(aes(
                x = cell_count,
                y = frip,
                shape = exp_type,
                color = exp_type,
                text = paste0("<b>Experiment Type</b>: ", exp_type, "<br>",
                                "<b>FRiP Score</b>: ", round(frip, 4),
                                "<br><b>Cell Count</b>: ", cell_count))) +
            geom_point() +
            geom_line(data = frip_df_cellcount) +
            labs(
                x = "Cell Count (log10 scale)",
                y = "FRiP Score",
                fill = "Experiment Type",
                shape = "Experiment Type",
                color = "Experiment Type"
            ) +
            scale_x_continuous(trans="log10") +
            scale_color_viridis(begin = 0.15, end = 0.6, discrete = TRUE,
                           option = "A")
        
        print(to_plotly(frip_cellcount_plot))
    }
}
```

## Peak Width Distribution {-}
This section presents the peak width distribution for each experiment type,
individual experiments and cell counts, reported in base pairs.

### By Experiment Type {- .unlisted}
```{r peakwidths_exp_plot1}
peak_width_plt1 <- peak_width_df %>%
    ggplot(aes(x = peak_width, fill = exp_type)) +
        geom_histogram(bins = 30, alpha = 0.5, position = "identity") +
        labs(
            x = "Peak Width",
            y = "Peak Count",
            fill = "Experiment Type"
        ) +
        scale_fill_viridis(begin = 0.2, end = 0.6, discrete = TRUE,
                            option = "A")
    
    print(to_plotly(peak_width_plt1))
```

```{r peakwidths_exp_plot2}
peak_width_plt2 <- peak_width_df %>%
    ggplot(aes(x = exp_type, y = peak_width, fill = exp_type)) +
        geom_boxplot(alpha = 0.5) +
        labs(
            x = "Experiment Type",
            y = "Peak Width",
            fill = "Experiment Type"
        ) +
        scale_fill_viridis(begin = 0.2, end = 0.6, discrete = TRUE,
                            option = "A")

    print(to_plotly(peak_width_plt2))
```

```{r peakwidths_exp_dt}
peak_width_df %>%
    group_by(exp_type) %>%
    rename("Experiment Type" = exp_type) %>%
    summarise(
        "Mean Peak Width" = round(mean(peak_width), 1),
        "Median Peak Width" = round(median(peak_width), 3),
        "2.5th Quantile" = round(quantile(peak_width, 0.025), 3),
        "97.5th Quantile" = round(quantile(peak_width, 0.975), 3),
        "Standard Deviation" = round(sd(peak_width), 1)
    ) %>%
    arrange(`Experiment Type`) %>%
    print_DT()
```

### By Individual Dataset {- .unlisted}
```{r peakwidths_ind_plot}
peak_width_plt3 <- peak_width_df %>%
    ggplot(aes(x = exp_label, y = peak_width, fill = exp_type)) +
        geom_boxplot(alpha = 0.5) +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        labs(
        x = "Experiment Label",
        y = "Peak Width",
        fill = "Experiment Type"
        ) +
        scale_fill_viridis(begin = 0.15, end = 0.6, discrete = TRUE,
                            option = "A")

    print(to_plotly(peak_width_plt3))
```

```{r peakwidths_ind_dt}
peak_width_df$cell_count <- rep(params$cell_counts, peak_width_len)

peak_width_df_all <- if (is.null(peak_width_df$cell_count)) {
    peak_width_df %>%
    group_by(exp_label, exp_type)
} else {
    peak_width_df %>%
    group_by(exp_label, exp_type, cell_count) %>%
    rename("Cell Count" = cell_count,) 
}

peak_width_df_all <- peak_width_df_all %>%
    summarise(
        "Mean Peak Width" = round(mean(peak_width), 1),
        "Median Peak Width" = round(median(peak_width), 3),
        "2.5th Quantile" = round(quantile(peak_width, 0.025), 3),
        "97.5th Quantile" = round(quantile(peak_width, 0.975), 3),
        "Standard Deviation" = round(sd(peak_width), 1),
        .groups = "drop"
    ) %>%
    rename(
        "Experiment Label" = exp_label,
        "Experiment Type" = exp_type
    ) %>%
    arrange(`Experiment Type`, `Experiment Label`)

    print_DT(peak_width_df_all)
```

```{r peakwidths_cellcount_plot}
if (!cellcount_metrics) {
    cat("\n", ex_emo,
    " *Skipping cell count plot as cell counts are not provided.*\n")
} else {
    cat("\n### By Cell Count {- .unlisted}  \n")
    
    peak_width_df_cellcount <- peak_width_df %>%
        group_by(exp_label, exp_type, cell_count) %>%
        drop_na(cell_count) %>%
        summarise(
            mean_width = mean(peak_width),
            .groups = "drop"
        ) %>%
        group_by(exp_type, cell_count) %>%
        summarise(
            mean_width = mean(mean_width),
            .groups = "drop"
        )
        
    peak_width_cellcount_plot <- peak_width_df %>%
        group_by(exp_label, exp_type) %>%
        drop_na(cell_count) %>%
        summarise(
            mean_width = mean(peak_width),
            cell_count = mean(cell_count),
            .groups = "drop"
        ) %>%
        ggplot(aes(
            x = cell_count,
            y = mean_width,
            shape = exp_type,
            color = exp_type,
            text = paste0("<b>Experiment Type</b>: ", exp_type, "<br>",
                            "<b>Mean Peak Width</b>: ", round(mean_width, 1),
                            "<br><b>Cell Count</b>: ", cell_count))) +
        geom_point() +
        geom_line(data = peak_width_df_cellcount) +
        labs(
            x = "Cell Count (log10 scale)",
            y = "Mean Peak Width",
            fill = "Experiment Type",
            shape = "Experiment Type",
            color = "Experiment Type"
        ) +
        scale_x_continuous(trans="log10") +
        scale_color_viridis(begin = 0.15, end = 0.6, discrete = TRUE,
                       option = "A")
    
    print(to_plotly(peak_width_cellcount_plot))
}
```

## Motif-Summit Distance {- .tabset .tabset-fade .tabset-pills}
This section reports the distances between the peak summit and the centre of the
nearest motif. The distances are calculated for every peak in the dataset and
plotted.
```{r motif_summit_dist}
if (!user_motif_metrics) {
    cat("\n", ex_emo,
    " *Skipping motif-summit distance plots as no motifs files",
    " are not provided.*\n")
} else {
    tab_header_flag <- FALSE # Print tabset header just once
    
    ## Loop through all motifs
    for (motif_i in seq_len(motif_len)) {
        if (!tab_header_flag) {
            cat("**Select Motif:**  \n")
            tab_header_flag <- TRUE
        }
        cat(paste0("\n### ", user_motifs$motifs[[motif_i]]@name,
                   " {- .unlisted}  \n"))
        cat(paste0("**Motif Consensus Sequence:** ",
                   user_motifs$motifs[[motif_i]]@consensus, "  \n"))
        
        cat("\n#### By Experiment Type {- .unlisted}  \n")
        cat("The top graph shows peak counts by distance from the motif",
        "center, while the bottom graph depicts the probability density of",
        "these distances. Different colors represent experiment types, and",
        "each density curve sums to 1.  \n")
        ## Raw count plot
        motif_summit_dist_plt1 <- motif_summit_dist_df %>%
            filter(motif_indice == motif_i) %>%
            ggplot(aes(x = distance, color = exp_type,
                        text = paste("<b>Experiment Type</b>:", exp_type))) +
            geom_freqpoly(bins = 30, linewidth = 1) +
            scale_color_viridis(begin = 0.2, end = 0.6, discrete = TRUE,
                            option = "A") +
            labs(color = "Experiment Type",
                x = "Distance (bp)", 
                y = "Count")
        motif_summit_dist_plt1 <- to_plotly(motif_summit_dist_plt1,
                                            html_tags = FALSE)
            
        ## Density plot
        motif_summit_dist_plt2 <- motif_summit_dist_df %>%
            filter(motif_indice == motif_i) %>%
            ggplot(aes(x = distance, fill = exp_type, color = exp_type,
                        text = paste("<b>Experiment Type</b>:", exp_type))) +
            geom_density(alpha = 0.5) +
            scale_fill_viridis(begin = 0.2, end = 0.6, discrete = TRUE,
                            option = "A") +
            scale_color_viridis(begin = 0.2, end = 0.6, discrete = TRUE,
                            option = "A") +
            labs(color = "Experiment Type",
                fill = "Experiment Type",
                x = "Distance (bp)", 
                y = "Count")
        motif_summit_dist_plt2 <- to_plotly(motif_summit_dist_plt2,
                                            html_tags = FALSE)
        
        motif_summit_dist_exp_plt <-
            subplot(motif_summit_dist_plt1, motif_summit_dist_plt2,
                    nrows = 2, margin = 0.04) %>%
            layout(
                xaxis2 = list(title = "Distance (bp)",
                              titlefont = list(size = 16)),
                yaxis = list(title = "Peak Count",
                             titlefont = list(size = 16)),
                yaxis2 = list(title = "Peak Density",
                              titlefont = list(size = 16))
            )
        print(htmltools::tagList(motif_summit_dist_exp_plt))
        
        ## DT
        motif_summit_dist_df %>%
            filter(motif_indice == motif_i) %>%
            group_by(exp_type) %>%
            summarise(
                "Mean Distance" = round(mean(distance), 2),
                "Absolute Mean Distance" = round(mean(abs(distance)), 2),
                "Median Distance" = median(distance),
                "Absolute Median Distance" = median(abs(distance)),
                "Standard Deviation" = round(sd(distance), 2)
            ) %>%
            print_DT(html_tags = TRUE) %>%
            print()
        
        cat("\n#### By Individual Dataset {- .unlisted}\n")
        cat("The plot shows the average of the absolute distance between peak",
        "summits and the nearest motif center for each experiment label.  \n")
        ## Absolute mean distance plot
        motif_summit_dist_ind_plt <- motif_summit_dist_df %>%
            filter(motif_indice == motif_i) %>%
            group_by(exp_label, exp_type) %>%
            summarise(abs_mean_dist = round(mean(abs(distance)), 2),
                      .groups = "drop") %>%
            ggplot(aes(x = exp_label, y = abs_mean_dist, fill = exp_type,
                        text = paste0(
                            "<b>Experiment Label</b>: ", exp_label, "<br>",
                            "<b>Experiment Type</b>: ", exp_type, "<br>",
                            "<b>Absolute Mean Distance</b>: ", abs_mean_dist
                        ))) +
                theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
                geom_col() +
                labs(
                  x = "Experiment Label",
                  y = "Absolute Mean Distance (bp)",
                  fill = "Experiment Type"
                ) +
            scale_fill_viridis(begin = 0.15, end = 0.6, discrete = TRUE,
                               option = "A")
        
            print(to_plotly(motif_summit_dist_ind_plt, html_tags = TRUE))
        
        ## DT
        motif_summit_dist_df %>%
            filter(motif_indice == motif_i) %>%
            group_by(exp_label, exp_type) %>%
            summarise(
                "Mean Distance" = round(mean(distance), 2),
                "Absolute Mean Distance" = round(mean(abs(distance)), 2),
                "Median Distance" = median(distance),
                "Absolute Median Distance" = median(abs(distance)),
                "Standard Deviation" = round(sd(distance), 2),
                .groups = "keep"
            ) %>%
            arrange(exp_type, exp_label) %>%
            rename(
                "Experiment Label" = exp_label,
                "Experiment Type" = exp_type
            ) %>%
            print_DT(html_tags = TRUE) %>%
            print()
    }
}
```

## Summary Table {-}
Summary of all the general metrics for the input datasets.

```{r summary_table}
tibble::tibble(
    "Experiment Label" = result$exp_label,
    "Experiment Type" = result$exp_type,
    "Cell Count" = params$cell_count,
    "Read Count" = result$read_count,
    "Peak Count" = result$peak_count,
    "FRiP Score" = round(frip_df$frip, 3),
    "Mean Peak Width" = peak_width_df_all$`Mean Peak Width`,
) %>%
    arrange(`Experiment Type`, `Experiment Label`) %>%
    print_DT(extra = TRUE)
```

# {- .unlisted}
<hr>  

# Known Motif Enrichment Analysis {.tabset .tabset-fade .tabset-pills #motif-enrichment-analysis}
Pair-wise comparison of peaks enriched for motifs in total, common and unique
peaks between the the reference experiment and others.  
  
The overall summary plot compacts all the individual comparison plots into one
interactive plot. It can be used to quickly identify the relative enrichment of
motifs in the peaks of the reference experiment compared to the other comparison
experiments.  
The sub-plots have been arranged in a 2x2 grid. The rows represent the peaks
from the reference experiment and the comparison experiments, respectively.
The columns differentiate between peaks that are common (shared) across
the comparison and reference experiments and those that are unique to either of
the experiments.  
  
The sub-plots for individual experiments can be interpreted as follows:  
  
1. `all_(experiment)_peaks`: Looks for the presence of motif in all the peaks
produced in an experiment.  
2. `common_(experiment)_peaks`: Looks for presence of motif in peaks which
overlap with the peaks from the other experiment. The overlapping peaks may be
of different lengths for each experiment, necessitating two distinct peak groups 
(one for each experiment).  
3. `unique_(experiment)_peaks`: Looks for presence of motif in peaks which have
no overlap with peaks from the other experiment.  

Each comparison compares the reference dataset with others.
  
**NOTE**: Discrepancy in motif enrichment in common peaks between experiments
MAY be attributed to different peak length.  

```{r motif_enrichment_analysis}
if (!user_motif_metrics) {
    cat("\n", ex_emo,
    " *Skipping motif-summit distance plots as no motifs files",
    " are not provided.*\n")
} else {
    tab_header_flag <- FALSE # Print tabset header just once
    
    ## Loop through all motifs
    for (motif_i in seq_len(motif_len)) {
        if (!tab_header_flag) {
            cat("**Select Motif:**  \n")
            tab_header_flag <- TRUE
        }
        cat(paste0("\n## ", user_motifs$motifs[[motif_i]]@name,
                   " {- .unlisted .tabset}  \n"))
        cat(paste0("**Motif Consensus Sequence:** ",
                   user_motifs$motifs[[motif_i]]@consensus, "  \n"))
        
        cat("<br>**Select Comparison Group:**  \n")
        
        ### Overall Summary Plot ###
        cat("\n### **Overall** {- .unlisted}  \n")
        cat("<details><summary>Show Comparison-Reference Pair Labels",
        "</summary>  \n")
        cat("(Pair Number. Comparison Experiment - Reference Experiment)  \n  ")
        for (pair_i in seq_along(comparison_indices)) {
          cat(paste0(
            pair_i, ". ", result$exp_labels[comparison_indices][pair_i], " - ",
            result$exp_labels[params$reference_index], "  \n"
          ))
        }
        cat("  \n</details>  \n")
        
        
        
        ### Individual Comparison Plots ###
        for (i in comparison_indices) {
            print_labels(result$exp_labels, params$reference_index, i,
                        result$read_count)
            
            ref_label <- result$exp_labels[params$reference_index]
            comp_label <- result$exp_labels[i]
            enrichment_df_filtered <- enrichment_df %>%
                filter(motif_indice == motif_i,
                        exp_label %in% c(ref_label, comp_label))
                
            enrichment_group_labels <- c(
                paste("All", ref_label, "Peaks"),
                paste("All", comp_label, "Peaks"),
                paste("Common", ref_label, "Peaks"),
                paste("Common", comp_label, "Peaks"),
                paste("Unique", ref_label, "Peaks"),
                paste("Unique", comp_label, "Peaks")
            )
            
            enrichment_group_labels_breaks <- c(
                paste("All\n", ref_label, "\nPeaks"),
                paste("All\n", comp_label, "\nPeaks"),
                paste("Common\n", ref_label, "\nPeaks"),
                paste("Common\n", comp_label, "\nPeaks"),
                paste("Unique\n", ref_label, "\nPeaks"),
                paste("Unique\n", comp_label, "\nPeaks")
            )
            
            ## Plot
            enrichment_df_filtered_plt <- enrichment_df_filtered %>%
                pivot_longer(cols = starts_with("count_") |
                                starts_with("perc_"),
                                    names_to = c(".value", "enriched"),
                                    names_sep = "_") %>%
                mutate(enriched = ifelse(enriched == "enriched",
                                        "Yes", "No")) %>%
                mutate(group = rep(enrichment_group_labels_breaks, each = 2)) %>%
                group_by(group) %>%
                ggplot(aes(x = group, y = count, fill = enriched,
                           text = paste0(
                               "<b>Peak Count:</b> ", count, " (", perc, "%)"
                           ))) +
                geom_bar(stat = "identity", position = "stack") +
                theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
                labs(
                    x = "Peak Groups",
                    y = "Peak Count"
                ) +
                scale_fill_manual(name = "Motif Present",
                                    labels = c("No", "Yes"),
                                    values = c("#fdae61", "#2b83ba"))
                
            enrichment_df_filtered_pltly <-
                to_plotly(enrichment_df_filtered_plt) %>%
                    layout(legend = list(traceorder = "reversed"))
            print(htmltools::tagList(enrichment_df_filtered_pltly))
            
            ## DT
            test_header <- htmltools::withTags(table(
                  class = 'display',
                  thead(
                    tr(
                      th(colspan = 2, "Category 1"),  # Span "Category 1" across 2 columns
                      th(colspan = 4, "Category 2")   # Span "Category 2" across 2 columns
                    ),
                    tr(
                      th(""), th("A"),               # Subheaders for "Category 1"
                      th("A"), th("B"), th("C"), th("D")                # Subheaders for "Category 2"
                    )
                  )
                ))
            
            cat("  \n  ")
            enrichment_DT <- enrichment_df_filtered %>%
                mutate(Group = enrichment_group_labels) %>%
                mutate(count_all = count_enriched + count_nonenriched,
                       .before = count_enriched) %>%
                select(`Group`, starts_with("count_"), perc_enriched) %>%
                rename(
                    "Total Peaks" = count_all,
                    "Peaks With Motif" = count_enriched,
                    "Peaks Without Motif" = count_nonenriched,
                    "Peaks With Motif (%)" = perc_enriched
                )
            
            print(print_DT(enrichment_DT, html_tags = TRUE))
            cat("  \n  ")
            
            
        }
    }
}
```

# {- .unlisted}
<hr>  

# Session Info {-}
<details> 
```{r Session Info, results='markup'}
utils::sessionInfo()
```
</details> 
<hr>
